libp2p/NodeTable.cpp

```diff
@@ -0,0 +1,466 @@
+/*
+ This file is part of cpp-ethereum.
+ 
+ cpp-ethereum is free software: you can redistribute it and/or modify
+ it under the terms of the GNU General Public License as published by
+ the Free Software Foundation, either version 3 of the License, or
+ (at your option) any later version.
+ 
+ cpp-ethereum is distributed in the hope that it will be useful,
+ but WITHOUT ANY WARRANTY; without even the implied warranty of
+ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ GNU General Public License for more details.
+ 
+ You should have received a copy of the GNU General Public License
+ along with cpp-ethereum.  If not, see <http://www.gnu.org/licenses/>.
+ */
+/** @file NodeTable.cpp
+ * @author Alex Leverington <nessence@gmail.com>
+ * @date 2014
+ */
+
+#include "NodeTable.h"
+using namespace std;
+using namespace dev;
+using namespace dev::p2p;
+
+NodeTable::NodeTable(ba::io_service& _io, KeyPair _alias, uint16_t _listenPort):
+	m_node(Node(_alias.pub(), bi::udp::endpoint())),
+	m_secret(_alias.sec()),
+	m_socket(new NodeSocket(_io, *this, _listenPort)),
+	m_socketPtr(m_socket.get()),
+	m_io(_io),
+	m_bucketRefreshTimer(m_io),
+	m_evictionCheckTimer(m_io)
+{
+	for (unsigned i = 0; i < s_bins; i++)
+	{
+		m_state[i].distance = i;
+		m_state[i].modified = chrono::steady_clock::now() - chrono::seconds(1);
+	}
+	
+	m_socketPtr->connect();
+	doRefreshBuckets(boost::system::error_code());
+}
+	
+NodeTable::~NodeTable()
+{
+	m_evictionCheckTimer.cancel();
+	m_bucketRefreshTimer.cancel();
+	m_socketPtr->disconnect();
+}
+
+void NodeTable::join()
+{
+	doFindNode(m_node.id);
+}
+	
+std::list<NodeId> NodeTable::nodes() const
+{
+	std::list<NodeId> nodes;
+	Guard l(x_nodes);
+	for (auto& i: m_nodes)
+		nodes.push_back(i.second->id);
+	return std::move(nodes);
+}
+
+list<NodeTable::NodeEntry> NodeTable::state() const
+{
+	list<NodeEntry> ret;
+	Guard l(x_state);
+	for (auto s: m_state)
+		for (auto n: s.nodes)
+			ret.push_back(*n.lock());
+	return move(ret);
+}
+
+NodeTable::NodeEntry NodeTable::operator[](NodeId _id)
+{
+	Guard l(x_nodes);
+	return *m_nodes[_id];
+}
+
+void NodeTable::requestNeighbors(NodeEntry const& _node, NodeId _target) const
+{
+	FindNode p(_node.endpoint.udp, _target);
+	p.sign(m_secret);
+	m_socketPtr->send(p);
+}
+
+void NodeTable::doFindNode(NodeId _node, unsigned _round, std::shared_ptr<std::set<std::shared_ptr<NodeEntry>>> _tried)
```

Comment:
Since you are `using namespace std` on top why not simnply omit `std::` in front of all STL stuff here and in other places to make the code more readable? I have to scroll to the right quite a bit to see the arguments of this function. Another idea is to split arguments into multiple lines.